### ISSUE TITLE
Create layout function for pyenv virtualenvwrapper

### ISSUE THREAD

**Issue by markfaine** [direnv/direnv#490]
## Issue description

I am using pyenv virtualenvwrapper and I'm trying to create a layout function that will issue workon to the environment when entering the cd and deactivate when exiting.  I also would like for it to create the venv if it doesn't already exist.  The issues I'm having is it that it would mean that I would have to load pyenv virtualenvwrapper or virtualenvwraper_lazy every time .envrc is read.  Is there some way to tell direnv to load this functions the first time it starts but not every time it starts:  This is what I have so far:

Does not work:
 eval "$(pyenv init -)"
    pyenv virtualenvwrapper_lazy
    layout_virtualenvwrapper() {
      project="$(basename "$1")"
      WORKON_HOME=~/.virtualenvs
      export WORKON_HOME
      PROJECT_HOME=~/projects
      export PROJECT_HOME
      if ! lsvirtualenv -b "${project}"; then
       mkvirtualenv "${project}"
     fi
     workon "$project"
   }

Sort of works, but doesn't run workon command.
    layout_virtualenvwrapper() {
      eval "$(pyenv init -)"
      project="$(basename "$1")"
      WORKON_HOME=~/.virtualenvs
      export WORKON_HOME
      PROJECT_HOME=~/projects
      export PROJECT_HOME
      if pyenv virtualenvwrapper_lazy; then
        if ! lsvirtualenv -b "${project}"; then
         mkvirtualenv "${project}"
       fi
     workon "$project"
     fi
   }

These are both already in my .bashrc so the best scenario would for them not to need to be referenced at all.  However, I think this is the part that is biting me:

> direnv is actually creating a new bash process to load the stdlib, direnvrc and .envrc, and only exports the environment diff back to the original shell. This allows direnv to record the environment changes accurately and also work with all sorts of shells. It also means that aliases and functions are not exportable right now.

### Steps to reproduce
Use the above function in .direnvrc and the following in the directory
.envrc:
layout_virtualenvwrapper "$(pwd)"

## Technical details

* target shell: bash
* direnv 2.7.0
* OS: Ubuntu 18.04

**Comment by zimbatm** [comment]
not really an answer but have you looked in the wiki yet? https://github.com/direnv/direnv/wiki/Python

**Comment by markfaine** [comment]
Yes, of course.  There seems to be layouts for a lot of python variations but I'm not sure how to make it work with [pyenv-virtualenvwrapper](https://github.com/pyenv/pyenv-virtualenvwrapper)

**Comment by zimbatm** [comment]
hopefully some python expert will come along, I don't have enough patience to learn yet another python packaging variation, sorry

**Comment by joshbode** [comment]
I'm just doing this in `~/.direnvrc`:
```
layout_workon() {
  eval "$(pyenv init -)"
  pyenv virtualenvwrapper
  workon "$1"
}
```
and putting this in `.envrc`:
```
layout workon .
```

which is working for me.

I think you just need to add the line `pyenv virtualwrapper` to your second attempt

**Comment by joshbode** [comment]
although, it works for me with `pyenv virtualenvwrapper_lazy` too

**Comment by wjewell3** [comment]
I was also struggling to marry up pyenv virtualenvwrapper with direnv. Here is my workaround:

1. My .direnvrc file is empty
2. I created a file called "envrc" (no dots) in Users/FYE7200/bin (which is on my path) with the following, so that once I git clone a repo, i can just type "envrc" and it does everything for me:

```
#!/bin/zsh
# deactivate this virtual environment bc it will be activated again as soon
# as direnv sees its `.envrc` file
bname=$(basename "$PWD")
source ${VIRTUALENVWRAPPER_SCRIPT}
mkvirtualenv ${bname}
source deactivate

# get name of this project
envrc_file=$(pwd)/.envrc

echo "Creating $envrc_file for direnv"
echo "
# create direnv's .envrc to activate ${bname}'s virtual environment
source ${VIRTUALENVWRAPPER_SCRIPT}
workon ${bname}
export PROJECT_HOME=$(pwd)
unset PS1
" >> $envrc_file

# tell direnv that newly created .envrc file is allowed
direnv allow
```
The output .envrc file looks like this:
```
# create direnv's .envrc to activate python-utilities's virtual environment
source /Users/FYE7200/.pyenv/versions/3.6.3/bin/virtualenvwrapper.sh
workon python-utilities
export PROJECT_HOME=/Users/FYE7200/Documents/Real_Time_Products/repos/python-utilities
unset PS1
```

Hope this helps!
